File: Startup.cs
using Microsoft.IdentityModel.Protocols.OpenIdConnect;
using Microsoft.Owin;
using Microsoft.Owin.Security;
using Microsoft.Owin.Security.Cookies;
using Microsoft.Owin.Security.Notifications;
using Microsoft.Owin.Security.OpenIdConnect;
using Okta.AspNet;
using Owin;
using System.Collections.Generic;
using System.Configuration;
using System.Security.Claims;
using System.Threading.Tasks;

[assembly: OwinStartup(typeof(ISAFIRRiskAssessmentWebApp.Startup))]

namespace ISAFIRRiskAssessmentWebApp
{
    public class Startup
    {
        public const string PreferredUsername = "preferred_username";
        public const string OktaUserId = "sub";
        public const string Roles = "roles";
        public const string AuthorizedAgencyEsam = "authorizedAgency";

        private readonly string _authority = ConfigurationManager.AppSettings["okta:OktaDomain"];
        private readonly string _clientId = ConfigurationManager.AppSettings["okta:ClientId"];
        private readonly string _clientSecret = ConfigurationManager.AppSettings["okta:ClientSecret"];
        private readonly string _redirectUri = ConfigurationManager.AppSettings["okta:RedirectUri"];

        public void Configuration(IAppBuilder app)
        {
            app.SetDefaultSignInAsAuthenticationType(CookieAuthenticationDefaults.AuthenticationType);

            app.UseCookieAuthentication(new CookieAuthenticationOptions());

            app.UseOktaMvc(new OktaMvcOptions()
            {
                OktaDomain = _authority,
                ClientId = _clientId,
                ClientSecret = _clientSecret,
                RedirectUri = _redirectUri,
                GetClaimsFromUserInfoEndpoint = true,
                Scope = new List<string> { "openid", "profile", "email" },
                OpenIdConnectEvents = new OpenIdConnectAuthenticationNotifications
                {
                    SecurityTokenValidated = OnTokenValidated
                }
            });
        }

        /// <summary>
        /// Custom logic to manipulate identity after token is validated.
        /// </summary>
        private Task OnTokenValidated(SecurityTokenValidatedNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions> context)
        {
            var currentIdentity = context.AuthenticationTicket.Identity;
            var customIdentity = new ClaimsIdentity(currentIdentity, new List<Claim>(), currentIdentity.AuthenticationType, PreferredUsername, Roles);
            var properties = context.AuthenticationTicket.Properties;
            var ticket = new AuthenticationTicket(customIdentity, properties);
            context.AuthenticationTicket = ticket;
            return Task.CompletedTask;
        }
    }
}

File: Startup.vb
Imports Microsoft.IdentityModel.Protocols.OpenIdConnect
Imports Microsoft.Owin
Imports Microsoft.Owin.Security
Imports Microsoft.Owin.Security.Cookies
Imports Microsoft.Owin.Security.Notifications
Imports Microsoft.Owin.Security.OpenIdConnect
Imports Okta.AspNet
Imports Owin
Imports System.Collections.Generic
Imports System.Configuration
Imports System.Security.Claims
Imports System.Threading.Tasks

<Assembly: OwinStartup(GetType(ISAFIRRiskAssessmentWebApp.Startup))>

Namespace ISAFIRRiskAssessmentWebApp
    Public Class Startup
        Public Const PreferredUsername As String = "preferred_username"
        Public Const OktaUserId As String = "sub"
        Public Const Roles As String = "roles"
        Public Const AuthorizedAgencyEsam As String = "authorizedAgency"

        Private ReadOnly _authority As String = ConfigurationManager.AppSettings("okta:OktaDomain")
        Private ReadOnly _clientId As String = ConfigurationManager.AppSettings("okta:ClientId")
        Private ReadOnly _clientSecret As String = ConfigurationManager.AppSettings("okta:ClientSecret")
        Private ReadOnly _redirectUri As String = ConfigurationManager.AppSettings("okta:RedirectUri")

        Public Sub Configuration(app As IAppBuilder)
            app.SetDefaultSignInAsAuthenticationType(CookieAuthenticationDefaults.AuthenticationType)

            app.UseCookieAuthentication(New CookieAuthenticationOptions())

            app.UseOktaMvc(New OktaMvcOptions() With {
                .OktaDomain = _authority,
                .ClientId = _clientId,
                .ClientSecret = _clientSecret,
                .RedirectUri = _redirectUri,
                .GetClaimsFromUserInfoEndpoint = True,
                .Scope = New List(Of String) From {"openid", "profile", "email"},
                .OpenIdConnectEvents = New OpenIdConnectAuthenticationNotifications() With {
                    .SecurityTokenValidated = AddressOf OnTokenValidated
                }
            })
        End Sub
        Private Function OnTokenValidated(context As SecurityTokenValidatedNotification(Of OpenIdConnectMessage, OpenIdConnectAuthenticationOptions)) As Task
            System.Diagnostics.Debug.WriteLine(">>> [Okta] Token validated")

            Dim currentIdentity = context.AuthenticationTicket.Identity
            Dim customIdentity = New ClaimsIdentity(currentIdentity, New List(Of Claim)(), currentIdentity.AuthenticationType, PreferredUsername, Roles)
            Dim properties = context.AuthenticationTicket.Properties
            Dim ticket = New AuthenticationTicket(customIdentity, properties)
            context.AuthenticationTicket = ticket

            Return Task.CompletedTask
        End Function

    End Class
End Namespace
